src/cursor/cur_version.c

```diff
@@ -0,0 +1,275 @@
+    hs_cursor = version_cursor->hs_cursor;
+    CURSOR_API_CALL(cursor, session, next, NULL);
+        
+    /* 
+     * - If there are untraversed updates in the update chain, return
+     * the next one. 
+     * - If there are no updates or all the updates have been exhausted,
+     * return the ondisk value.
+     * - Once the ondisk value is exhausted, check if there are any values
+     * in the history store.
+     */
+    upd = version_cursor->next_upd;
+    if (upd != NULL && !F_ISSET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED)) {
+        upd = upd->next;
+        
+        if (upd == NULL)
+            F_SET(version_cursor, WT_VERSION_CUR_UPDATE_EXHAUSTED);
+    } else if (!F_ISSET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED)) {
+        F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+    } else if (!F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /* Use the history store cursor to position on the key. */
+        hs_cursor->next(hs_cursor);
+
+        if (ret != 0)
+            F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+    } else {
+        /* We have exhausted all versions of the key. */
+        ret = WT_NOTFOUND;
+    }
+
+    if (0) {
+err:
+        WT_TRET(cursor->reset(cursor));
+    }
+    API_END_RET(session, ret);
+    return (0);
+}
+
+/*
+ * __curversion_reset --
+ *     WT_CURSOR::reset for version cursors.
+ */
+static int
+__curversion_reset(WT_CURSOR *cursor)
+{
+    WT_CURSOR *hs_cursor, *table_cursor;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    hs_cursor = version_cursor->hs_cursor;
+    table_cursor = version_cursor->table_cursor;
+    CURSOR_API_CALL(cursor, session, reset, NULL);
+
+    if (table_cursor != NULL)
+        WT_TRET(table_cursor->reset(table_cursor));
+    if (hs_cursor != NULL)
+        WT_TRET(hs_cursor->reset(hs_cursor));
+    version_cursor->next_upd = NULL;
+    version_cursor->flags = 0;
+    F_CLR(cursor, WT_CURSTD_KEY_SET);
+    F_CLR(cursor, WT_CURSTD_VALUE_SET);
+
+err:
+    API_END_RET(session, ret);
+}
+
+/*
+ * __curversion_search --
+ *     WT_CURSOR->search method for version cursors.
+ */
+static int
+__curversion_search(WT_CURSOR *cursor)
+{
+    WT_CURSOR *table_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_VERSION *version_cursor;
+    WT_DECL_RET;
+    WT_INSERT *ins;
+    WT_PAGE *page;
+    WT_ROW *rip;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *first_upd, *upd;
+    bool key_only;
+
+    version_cursor = (WT_CURSOR_VERSION *)cursor;
+    table_cursor = version_cursor->table_cursor;
+    key_only = F_ISSET(cursor, WT_CURSTD_KEY_ONLY);
+
+    /*
+     * For now, we assume that we are using simple cursors only, and row store only.
+     */
+    cbt = (WT_CURSOR_BTREE *)table_cursor;
+    CURSOR_API_CALL(cursor, session, search, CUR2BT(cbt));
+    WT_ERR(__cursor_checkkey(cursor));
```

Comment:
This is wrong, you should check table cursor instead as the set key function directly sets the key in table cursor not the key on the version cursor.